pkg/virt-controller/watch/vmi.go

```diff
@@ -353,6 +353,67 @@ func (c *VMIController) execute(key string) error {
 	return nil
 }
 
+// These "dynamic" labels are Pod labels which may diverge from the VMI over time that we want to keep in sync.
+func (c *VMIController) syncDynamicLabelsToPod(pod *k8sv1.Pod, vmi *virtv1.VirtualMachineInstance) error {
+	var patchOps []string
+
+	dynamicLabels := []string{
+		virtv1.NodeNameLabel,
+		virtv1.OutdatedLauncherImageLabel,
+	}
+
+	newMeta := pod.ObjectMeta.DeepCopy()
+	if newMeta.Labels == nil {
+		newMeta.Labels = map[string]string{}
+	}
+
+	changed := false
+	for _, key := range dynamicLabels {
+		val, exists := vmi.Labels[key]
+		podVal, podLabelExists := newMeta.Labels[key]
+		if exists == podLabelExists && val == podVal {
+			continue
+		}
+
+		changed = true
+		if !exists {
+			delete(newMeta.Labels, key)
+		} else {
+			newMeta.Labels[key] = val
+		}
+	}
+
+	if !changed {
+		return nil
+	}
+
+	newLabelBytes, err := json.Marshal(newMeta.Labels)
+	if err != nil {
+		return err
+	}
+	if pod.ObjectMeta.Labels == nil {
+		patchOps = append(patchOps, fmt.Sprintf(`{ "op": "add", "path": "/metadata/labels", "value": %s }`, string(newLabelBytes)))
+	} else {
+		oldLabelBytes, err := json.Marshal(pod.ObjectMeta.Labels)
+		if err != nil {
+			return err
+		}
+		patchOps = append(patchOps, fmt.Sprintf(`{ "op": "test", "path": "/metadata/labels", "value": %s }`, string(oldLabelBytes)))
+		patchOps = append(patchOps, fmt.Sprintf(`{ "op": "replace", "path": "/metadata/labels", "value": %s }`, string(newLabelBytes)))
+	}
```

Comment:
kubevirt project jas json-patch at the go.mod so we can use it instead of harcoding the pos

https://github.com/evanphx/json-patch

Reply:
i'm not sure how json-patch helps us here. We aren't applying a patch to a struct. We're passing the patch to the api server to apply.